File: frontend/src/api/services/pricingPolicyService.ts
import { api } from "@/api/axiosInstance";
import { API_ENDPOINTS } from "@/api/endpoints";
import type { ApiResponse } from "@/types/api";
import type {
  PricingPolicy,
  CreatePricingPolicyRequest,
  UpdatePricingPolicyRequest
} from "@/types/pricingPolicy";
import { executeApiRequest } from "./serviceExecutor";

export const pricingPolicyService = {
  getAll: (): Promise<ApiResponse<PricingPolicy[]>> => {
    return executeApiRequest<PricingPolicy[]>(() =>
      api.get(API_ENDPOINTS.PRICING_POLICY.GET_ALL),
    );
  },

  getById: (id: string): Promise<ApiResponse<PricingPolicy>> => {
    return executeApiRequest<PricingPolicy>(() =>
      api.get(API_ENDPOINTS.PRICING_POLICY.GET_DETAIL(id)),
    );
  },

  create: (payload: CreatePricingPolicyRequest): Promise<ApiResponse<string>> => {
    return executeApiRequest<string>(() =>
      api.post(API_ENDPOINTS.PRICING_POLICY.CREATE, payload),
    );
  },

  update: (payload: UpdatePricingPolicyRequest): Promise<ApiResponse<void>> => {
    return executeApiRequest<void>(() =>
      api.put(API_ENDPOINTS.PRICING_POLICY.UPDATE, payload),
    );
  },

  delete: (id: string): Promise<ApiResponse<void>> => {
    return executeApiRequest<void>(() =>
      api.delete(API_ENDPOINTS.PRICING_POLICY.DELETE(id)),
    );
  },

  setAsDefault: (id: string): Promise<ApiResponse<void>> => {
    return executeApiRequest<void>(() =>
      api.put(API_ENDPOINTS.PRICING_POLICY.SET_DEFAULT(id)),
    );
  },
};

File: src/Application/Services/PricingPolicyService.cs
using Application.Common.Constant;
using Application.Contracts.PricingPolicy;
using Contracts;
using Domain.Common.Repositories;
using Domain.Entities;
using Domain.Enums;
using Domain.UnitOfWork;
using ErrorOr;

namespace Application.Services;

public interface IPricingPolicyService
{
    Task<ErrorOr<List<PricingPolicyResponse>>> GetAll();
    Task<ErrorOr<PricingPolicyResponse>> GetById(Guid id);
    Task<ErrorOr<Guid>> Create(CreatePricingPolicyRequest request);
    Task<ErrorOr<Success>> Update(UpdatePricingPolicyRequest request);
    Task<ErrorOr<Success>> Delete(Guid id);
    Task<ErrorOr<Success>> SetAsDefault(Guid id);
}

public class PricingPolicyService(
    IPricingPolicyRepository pricingPolicyRepository,
    IUnitOfWork unitOfWork) : IPricingPolicyService
{
    private readonly IPricingPolicyRepository _pricingPolicyRepository = pricingPolicyRepository;
    private readonly IUnitOfWork _unitOfWork = unitOfWork;

    public async Task<ErrorOr<List<PricingPolicyResponse>>> GetAll()
    {
        var policies = await _pricingPolicyRepository.FindAll();
        return policies.Select(MapToResponse).ToList();
    }

    public async Task<ErrorOr<PricingPolicyResponse>> GetById(Guid id)
    {
        var policy = await _pricingPolicyRepository.FindById(id);
        if (policy is null)
            return Error.NotFound(ErrorConstants.PricingPolicy.NotFoundCode, ErrorConstants.PricingPolicy.NotFoundDescription);

        return MapToResponse(policy);
    }

    public async Task<ErrorOr<Guid>> Create(CreatePricingPolicyRequest request)
    {
        var policy = PricingPolicy.Create(
            request.Name,
            request.TourType,
            request.Tiers,
            request.IsDefault,
            "system",
            request.Translations);

        if (request.IsDefault)
        {
            await _pricingPolicyRepository.RemoveDefaultFromOthers(policy.Id);
        }

        await _pricingPolicyRepository.Create(policy);
        await _unitOfWork.SaveChangeAsync();
        return policy.Id;
    }

    public async Task<ErrorOr<Success>> Update(UpdatePricingPolicyRequest request)
    {
        var policy = await _pricingPolicyRepository.FindById(request.Id);
        if (policy is null)
            return Error.NotFound(ErrorConstants.PricingPolicy.NotFoundCode, ErrorConstants.PricingPolicy.NotFoundDescription);

        policy.Update(request.Name, request.TourType, request.Tiers, "system");

        // Update status if provided
        if (request.Status.HasValue)
        {
            policy.SetStatus(request.Status.Value, "system");
        }

        // Update translations if provided
        if (request.Translations != null)
        {
            policy.Translations = request.Translations;
        }

        await _pricingPolicyRepository.UpdateAsync(policy);
        await _unitOfWork.SaveChangeAsync();
        return Result.Success;
    }

    public async Task<ErrorOr<Success>> Delete(Guid id)
    {
        var policy = await _pricingPolicyRepository.FindById(id);
        if (policy is null)
            return Error.NotFound(ErrorConstants.PricingPolicy.NotFoundCode, ErrorConstants.PricingPolicy.NotFoundDescription);

        policy.SoftDelete("system");
        await _pricingPolicyRepository.UpdateAsync(policy);
        await _unitOfWork.SaveChangeAsync();
        return Result.Success;
    }

    public async Task<ErrorOr<Success>> SetAsDefault(Guid id)
    {
        var policy = await _pricingPolicyRepository.FindById(id);
        if (policy is null)
            return Error.NotFound(ErrorConstants.PricingPolicy.NotFoundCode, ErrorConstants.PricingPolicy.NotFoundDescription);

        await _pricingPolicyRepository.RemoveDefaultFromOthers(id);
        await _pricingPolicyRepository.SetAsDefault(id);
        await _unitOfWork.SaveChangeAsync();
        return Result.Success;
    }

    private static PricingPolicyResponse MapToResponse(PricingPolicy policy)
    {
        return new PricingPolicyResponse(
            policy.Id,
            policy.PolicyCode,
            policy.Name,
            policy.TourType,
            policy.TourType.ToString(),
            policy.Status,
            policy.Status.ToString(),
            policy.IsDefault,
            policy.Tiers,
            policy.Translations ?? [],
            policy.CreatedOnUtc,
            policy.LastModifiedOnUtc);
    }
}
